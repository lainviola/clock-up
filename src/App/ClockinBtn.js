// @flow
import React, { Component } from 'react'
import { graphql, compose } from 'react-apollo'
import styled from 'styled-components'
import gql from 'graphql-tag'
import { fetchUserQuery } from '../graphql'
import { Button } from '../components/Button'

type Props = {
  data: Object,
  mutation1: Function,
  mutation2: Function
}

export class ClockinBtn extends Component<Props> {
  constructor(props: Props) {
    super(props)
    this.recordClockinTimeToGraphcool = this.recordClockinTimeToGraphcool.bind(
      this
    )
  }

  recordClockinTimeToGraphcool() {
    const { data, mutation1, mutation2 } = this.props

    const userId = data.user.id
    const clockIn = () => new Date().toISOString()

    mutation1({
      variables: { userId },
      refetchQueries: [{ query: fetchUserQuery }]
    }).then(response => {
      console.log(response)
    })

    mutation2({
      variables: { userId: userId, clockIn: clockIn() }
    }).then(response => {
      console.log(response)
    })
  }

  render() {
    return (
      <MyButton
        onClick={this.recordClockinTimeToGraphcool}
        data-test="clock-in-btn"
      >
        Clock In
      </MyButton>
    )
  }
}

const MyButton = Button.extend`
  color: #ffffff;
  background-color: #66bb6a;
  border: 2px solid #66bb6a;
`

const mutation1 = gql`
  mutation($userId: ID!) {
    updateUser(id: $userId, isDuringClockIn: true) {
      isDuringClockIn
    }
  }
`

const mutation2 = gql`
  mutation($userId: ID!, $clockIn: DateTime) {
    createClock(userId: $userId, clockIn: $clockIn) {
      id
      clockIn
      clockOut
    }
  }
`

export default compose(
  graphql(mutation1, {
    name: 'mutation1'
  }),
  graphql(mutation2, {
    name: 'mutation2'
  })
)(ClockinBtn)
