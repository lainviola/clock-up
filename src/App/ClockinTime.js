// @flow
import React, { Component } from 'react'
import { connect } from 'react-redux'
import styled from 'styled-components'
import { borderColor, color } from '../cssVariables'

type Props = {
  data: Object,
  dateObject: Date
}

export class ClockinTime extends Component<Props> {
  constructor(props: Props) {
    super(props)
    this.formatDate = this.formatDate.bind(this)
  }

  /**
   * @param str ISOString. e.g. '2017-12-10T14:31:10.501Z'
   * @returns {string} human readable string. e.g. '11:31 PM'
   */
  formatDate(str: string): string {
    const dateObj = new Date(str)

    const options = {
      hour: '2-digit',
      minute: '2-digit'
    }

    return dateObj.toLocaleTimeString('en-us', options)
  }

  getWorkingTime(latestClockinTime: string): string {
    const { dateObject } = this.props

    const clockin = new Date(latestClockinTime)
    const now = dateObject
    const diff = now - clockin // milliseconds
    var msec = diff
    const hh = Math.floor(msec / 1000 / 60 / 60)
    msec -= hh * 1000 * 60 * 60
    const mm = Math.floor(msec / 1000 / 60)
    msec -= mm * 1000 * 60
    const ss = Math.floor(msec / 1000)
    msec -= ss * 1000

    return hh + 'h' + mm + 'm' + ss + 's'
  }

  render() {
    const { data } = this.props

    const latestClockinTime: string = data.user.clocks[0].clockIn
    return (
      <Container data-test="clock-in-time">
        <Text>ClockIn</Text>
        <Text>{this.formatDate(latestClockinTime)}</Text>
        <Text>WoringTime</Text>
        <Text>{this.getWorkingTime(latestClockinTime)}</Text>
      </Container>
    )
  }
}

const Container = styled.div`
  flex-grow: 3;
  color: ${color};
  font-size: 1.1em;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-around;
  border: 1px solid ${borderColor};
  border-radius: 5px;
  margin: 0 10px;
`

const Text = styled.div``

function mapStateToProps(state) {
  return {
    dateObject: state.app.dateObject
  }
}

export default connect(mapStateToProps)(ClockinTime)
