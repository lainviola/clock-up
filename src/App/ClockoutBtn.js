// @flow
import React, { Component } from 'react'
import { graphql, compose } from 'react-apollo'
import styled from 'styled-components'
import gql from 'graphql-tag'
import { fetchUserQuery } from '../graphql'

type Props = {
  data: Object,
  mutation1: Function,
  mutation2: Function
}

export class ClockoutBtn extends Component<Props> {
  constructor(props: Props) {
    super(props)
    this.recordClockoutTimeToGraphcool = this.recordClockoutTimeToGraphcool.bind(
      this
    )
  }

  recordClockoutTimeToGraphcool() {
    const { data, mutation1, mutation2 } = this.props

    const userId = data.user.id
    const clockOut = () => new Date().toISOString()

    mutation1({
      variables: { userId },
      refetchQueries: [{ query: fetchUserQuery }]
    }).then(response => {
      // TODO console.log()
      console.log(response)
    })

    mutation2({
      variables: {
        clockId: data.user.clocks[0].id,
        userId: userId,
        clockOut: clockOut()
      }
    }).then(response => {
      // TODO console.log()
      console.log(response)
    })
  }

  render() {
    return (
      <Button
        onClick={this.recordClockoutTimeToGraphcool}
        data-test="clock-out-btn"
      >
        clock out
      </Button>
    )
  }
}

const Button = styled.div`
  width: 70px;
  color: #ffffff;
  font-size: 1em;
  background-color: #e57373;
  line-height: 28px;
  padding: 0.25em 1em;
  border: 2px solid #e57373;
  border-radius: 3px;
  cursor: pointer;
`

const mutation1 = gql`
  mutation($userId: ID!) {
    updateUser(id: $userId, isDuringClockIn: false) {
      isDuringClockIn
    }
  }
`

const mutation2 = gql`
  mutation($clockId: ID!, $userId: ID, $clockOut: DateTime) {
    updateClock(id: $clockId, userId: $userId, clockOut: $clockOut) {
      id
      clockIn
      clockOut
    }
  }
`

export default compose(
  graphql(mutation1, {
    name: 'mutation1'
  }),
  graphql(mutation2, {
    name: 'mutation2'
  })
)(ClockoutBtn)
