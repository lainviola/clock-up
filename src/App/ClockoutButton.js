// @flow
import React, { Component } from 'react'
import { graphql, compose } from 'react-apollo'
import gql from 'graphql-tag'
import { fetchUserQuery } from '../graphql'
import { Button } from '../components/Button'

type Props = {
  data: Object,
  mutation1: Function,
  mutation2: Function
}

export class ClockoutButton extends Component<Props> {
  constructor(props: Props) {
    super(props)
    this.recordClockoutTimeToGraphcool = this.recordClockoutTimeToGraphcool.bind(
      this
    )
  }

  recordClockoutTimeToGraphcool() {
    const { data, mutation1, mutation2 } = this.props

    const userId = data.user.id
    const clockOut = () => new Date().toISOString()

    mutation1({
      variables: { userId },
      refetchQueries: [{ query: fetchUserQuery }]
    }).then(response => {
      // TODO console.log()
      console.log(response)
    })

    mutation2({
      variables: {
        clockId: data.user.clocks[0].id,
        userId: userId,
        clockOut: clockOut()
      }
    }).then(response => {
      // TODO console.log()
      console.log(response)
    })
  }

  render() {
    return (
      <MyButton
        onClick={this.recordClockoutTimeToGraphcool}
        data-test="clock-out-btn"
      >
        clock out
      </MyButton>
    )
  }
}

const MyButton = Button.extend`
  color: #ffffff;
  background-color: #e57373;
  border: 2px solid #e57373;
`

const mutation1 = gql`
  mutation($userId: ID!) {
    updateUser(id: $userId, isDuringClockIn: false) {
      isDuringClockIn
    }
  }
`

const mutation2 = gql`
  mutation($clockId: ID!, $userId: ID, $clockOut: DateTime) {
    updateClock(id: $clockId, userId: $userId, clockOut: $clockOut) {
      id
      clockIn
      clockOut
    }
  }
`

export default compose(
  graphql(mutation1, {
    name: 'mutation1'
  }),
  graphql(mutation2, {
    name: 'mutation2'
  })
)(ClockoutButton)
