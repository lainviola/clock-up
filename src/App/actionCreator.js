import { parseTime } from '../common/util'
import type { CurrentTime } from '../types/CurrentTime'
import { SYNC_DATE } from '../common/const'
import type { SyncDateAction } from '../types/ReduxAction'

export function syncDate(): SyncDateAction {
  const time: CurrentTime = parseTime(new Date())
  return {
    type: SYNC_DATE,
    currentTime: time
  }
}
