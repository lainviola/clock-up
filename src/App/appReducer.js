// @flow
import { parseTime } from '../common/util'
import type { CurrentTime } from '../types/CurrentTime'
import type { AppReduxState } from '../types/AppReduxState'
import type { ReduxAction } from '../types/ReduxAction'
import type { HistoryQueryParameter } from '../types/HistoryQueryParameter'

const currentTime: CurrentTime = parseTime(new Date())
const historyQueryParameter: HistoryQueryParameter = {
  first: 7,
  orderBy: 'createdAt_DESC'
}

const initialState: AppReduxState = {
  currentTime: currentTime,
  historyQueryParameter: historyQueryParameter
}

export default function appReducer(
  state: AppReduxState = initialState,
  action: ReduxAction
): AppReduxState {
  switch (action.type) {
    case 'SYNC_DATE':
      const currentTime: CurrentTime = action.currentTime

      return Object.assign({}, state, { currentTime: currentTime })
    default:
      return state
  }
}
