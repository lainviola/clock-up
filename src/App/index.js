// @flow
import React, { Component } from 'react'
import { withRouter } from 'react-router-dom'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'
import LoginAuth0 from './LoginAuth0'
import { AUTH0_ID_TOKEN } from '../GlobalConst'

type Props = {
  data: any
}

class App extends Component<Props> {
  _logout = () => {
    window.localStorage.removeItem(AUTH0_ID_TOKEN)
    window.location.reload()
  }

  _isLoggedIn = () => {
    return this.props.data.user
  }

  render() {
    if (this.props.data.loding) {
      return <div>Loading</div>
    }

    if (this._isLoggedIn()) {
      return this.renderLoggedIn()
    } else {
      return this.renderAuth0()
    }
  }

  renderLoggedIn() {
    return (
      <div>
        <span onClick={this._logout}>Logout</span>
      </div>
    )
  }

  renderAuth0() {
    return <LoginAuth0 />
  }
}

const userQuery = gql`
  query {
    user {
      id
    }
  }
`

export default graphql(userQuery, { options: { fetchPolicy: 'network-only' } })(
  withRouter(App)
)
