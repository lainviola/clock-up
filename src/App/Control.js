// @flow
import React, { Component } from 'react'
import styled from 'styled-components'
import ClockinBtn from './ClockinButton'
import ClockinTime from './ClockinTime'
import ClockoutBtn from './ClockoutButton'

type Props = {
  data: Object
}

export class Control extends Component<Props> {
  render() {
    const { data } = this.props

    if (data.user.isDuringClockIn) {
      return (
        <ClockoutContainer>
          <ClockinTime data={data} />
          <ClockoutBtn data={data} />
        </ClockoutContainer>
      )
    } else {
      return (
        <ClockinContainer>
          <ClockinBtn data={data} />
        </ClockinContainer>
      )
    }
  }
}

const ClockoutContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  padding: 0 20px;
`

const ClockinContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
  padding: 0 20px;
`

export default Control
