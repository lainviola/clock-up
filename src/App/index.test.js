import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import App from './index'
import { apolloWarapper } from '../testUtil'

describe('Component: App.index', () => {
  it('snapshot', () => {
    const wrapper = shallow(apolloWarapper(App))
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
