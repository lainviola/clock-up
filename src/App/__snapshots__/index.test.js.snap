// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: App.index snapshot 1`] = `
<QueryRecyclerProvider>
  <Apollo(withRouter(App)) />
</QueryRecyclerProvider>
`;
